presc/misclassifications/misclass_rate.py

```diff
@@ -50,25 +51,52 @@ def misclass_rate_feature(
             of bins, this parameter allows to specify whether these bins should
             be "regular" evenly spaced bins or "quantiles". Default value is
             "regular".
+        metric_function(lambda function): This parameter will take in a lambda function
+            which will work on matched true y value and predicted y value. Please make sure
+            to write a runnable metric score function(from sklearn.metric) with all parameters
+            (with its value) you want to set. This misclass_rate_feature function will work
+            exactly the metric function which you give on binned data.
+            Such as: metric_function=lambda y_true, y_pred: f1_score(y_true, y_pred, pos_label="True", zero_division=0)
 
     Returns:
         Three elements which correspond to:
             1) The edges of the bins in the feature scale.
             2) The misclassification rate in each bin.
             3) The standard deviation of the misclassification rate in that bin.
+            4) The list of conditional metrics in each bin.
     """
     # Builds dataset with only the misclassified data points
     test_dataset_misclass = test_dataset[test_dataset.iloc[:, -1] != test_predictions]
 
+    y_test_name = list(test_dataset.columns)[-1]
+
+    y_pred_series = pd.Series(test_predictions, test_dataset.index)
+
     if categorical is False:
 
-        # Computes position of bin edges for quartiles or deciles
+        # Computes position of bin edges and groups for quartiles or deciles
         if bins == "quartiles":
             bins = compute_quantiles(test_dataset, feature, quantiles=4)
+            groups = test_dataset.groupby(
+                by=pd.qcut(test_dataset[feature], q=4, duplicates="drop")
+            )
+
         elif bins == "deciles":
             bins = compute_quantiles(test_dataset, feature, quantiles=10)
+            groups = test_dataset.groupby(
+                by=pd.qcut(test_dataset[feature], q=10, duplicates="drop")
+            )
+
         elif type(bins) == int and bins_type == "quantiles":
+            bin_num = bins
             bins = compute_quantiles(test_dataset, feature, quantiles=bins)
+            groups = test_dataset.groupby(
+                by=pd.qcut(test_dataset[feature], q=bin_num, duplicates="drop")
+            )
+
+        # groups for regular bin edges or known bin edges
+        else:
+            groups = test_dataset.groupby(by=pd.cut(test_dataset[feature], bins=bins))
 
         # Histogram of all points
         total_histogram_counts, bins = np.histogram(test_dataset[feature], bins)
```

Comment:
You can now remove the computations involving `total_histogram_counts` and `misclass_histogram_counts`, since the function now computes more general metrics than just misclassification rates.

Reply:
Removed that.